README.md
## DevOps: A Continuous Delivery Environment for Automatically Building and Deploying Java Projects using Maven

Version 2.0.0 16th March 2019

This project contains a build script to create a DevOps continuous delivery server/system. This will enable you to automatically build a complete working continuous delivery environment from native installations and Docker images to support your Java development projects built using Maven. 

The script is written for installation on Centos 7.

The environment has the following components:

1. GitBlit: A lightweight Git repository manager to store the source code for your project.
2. Maven: I'm using Maven for code compilation and dependecy management.
3. Jenkins: This is the task runner for the environment. Create jobs in Jenkins to poll your Git repo for changes and then trigger a build automatically.
4. Nexus: For storing built artifacts and dependencies. Proxies Maven Central for seemless downloading of dependencies.
5. SonarQube: To analyse your code to identify code smells, bugs and vulnerabilities. 
6. This project installs Docker, Jenkins and Maven natively, and then uses Docker images of all the other tools.

### Workflow

OK, here is the workflow. You check your code into GitBlit throughout the day as you develop it. You configure Jenkins to poll GitBlit at regular intervals checking for new commits. On finding one it will checkout the code and build it getting any dependencies from the Nexus respository (that also proxies Maven Central for convinience). It will identify any build issues and once successfully built with the 'package' Maven goal, it will run your automated tests to check for any failures. If your tests pass, the final step is to analyse your code using SonarQube. You can then view the results of the analysis and rectify any problems early in the development lifecycle.

The idea behind this project is simple, develop good quality code and detect issues early on when they can be easily fixed.

Semantic versioning is applied to this project.

Good luck and enjoy !

### Instructions to build the environment

1. First you need to create a basic Linux server by installing Centos 7. This can be any type of server, physical, VM or cloud based.

2. Install 'git' so that you can download this repository

```ruby
yum install -y git
```

3. Clone this repository onto your new Centos 7 server into the directory of your choice by replacing {your-directory}
```ruby
git clone https://github.com/DocsWebApps/devops-jhipster-continuous-delivery-build-stack.git {your-directory} 
```

4. Set the environment variables to your choice in the file set_env.bash. This involves selecting where you want the container bind mounts, what ports your want each service in the environment to listen on and which versions of each component do you wish to use. You can also set up the hostname of your server here.  

5. Run the build script as root, once completed you should have a fully built CI environment with all the tools installed and ready to start up.

```ruby
./build.bash
``` 

6. Now you can start your stack using the command below. This will start Jenkins and download all the Docker images and then start all the Docker containers. You can check the status of your containers using the command 'docker container ls'. 

```ruby
./start_stack.bash
``` 

7. Now you need to configure the environment by following the instructions below

### Instructions to configure the environment

#### GitBlit
1. Naviagte to http://{serverhostname}:{port: default 9010} 

2. You can now login as 'admin' with the password 'admin' and create other users and teams as you please. 

4. For more information about GitLab and how to use it please visit: <a href="http://gitblit.com/" target="_blank">GitBlit</a>

#### Jenkins
1. Naviagte to http://{serverhostname}:8080

2. You will be asked to unlock Jenkins by entering the admin password. This can be found in the file below. Simply 'cat' this file and copy the contents into the Admistrator password field in the browser. 

```ruby
${DOCKER_VOLUMES}/secrets/initialAdminPassword
```

3. Jenkins will then ask you if you want to 'Install suggested plugins' - choose this option and let Jenkins download and install them

4. Once the plugins have been installed, you can create your first admin user

5. Jenkins is now ready to use, but you'll need to download and install more plugins as required for your projects

6. More information about how to setup and use Jenkins can be found at: <a href="https://jenkins.io/" target="_blank">Jenkins</a>

#### Nexus
1. Navigate to http://{serverhostname}:{port: default 9050}/nexus/#welcome

2. Login using the default username / password of: admin / admin123. If you choose to change the 'admin' password now, please read the notes below in the section 'Maven/Nexus'.

3. Setup new users and repositories as you need

4. For more information about Nexus, please visit: <a href="https://support.sonatype.com/hc/en-us/categories/201980798" target="_blank">Nexus</a>

#### SonarQube
1. Navigate to http://{serverhostname}:{port: default 9060}

2. Login using the default username / password of: admin / admin. Now change the 'admin' password.

3. You will be asked if you want to setup a token, but you can skip this tutorial

4. For more information about SonarQube, please visit: <a href="https://www.sonarqube.org/" target="_blank">SonarQube</a>


#### Maven/Nexus Dependency Management
I have configured Maven to download dependencies to its own local respository stored in '${BASE_DIR}/maven/maven-repo' via the Nexus artifact repository. This enables you to build up an artifact repository in Nexus that can then be used by other servers/services and/or developers to retrieve dependencies. Maven connects to the Nexus repository server via the configuration setup in the '${$MAVEN_DIR}/mvn3/conf/settings.xml' file. It uses the Nexus admin username/password you setup in the 'set_env.bash' file, but you can change these settings easily by changing the admin password in Nexus, then modify the NEXUS_PASS environment variable in the file 'set_env.bash', and then finally updating the password in the Maven settings file.

### Starting and stopping your environment

You can start up or stop your environment by simply using the following commands:

```ruby
./start_stack.bash - to start the environment
```

```ruby
./stop_stack.bash - to stop the environment
```

### License

This project is released under the <a href="http://www.opensource.org/licenses/MIT" target="_blank">MIT License</a>.


CHANGELOG.md
Bug Fixes / Niddles / Enhancements
----------------------------------

Future Features
---------------
1. Add a build for Ubuntu

Release v2.0.0 16th March 2019 
------------------------------

Release v1.1.0 29th August 2018 
-------------------------------
* 2e0fd90 (HEAD, tag: v1.1.0, origin/master, master) Moved away from GitLab and using GitBlit - 29/08/2018

Release v1.0.0 25th July 2018 
-----------------------------
* 5936635 (tag: v1.0.0) RC1 candidate - 25/07/2018
* 8cf9d4b RC1 candidate - 25/07/2018
* 2390728 RC1 candidate - 25/07/2018
* 9390829 RC1 candidate - 25/07/2018
* fccf74e RC1 candidate - 25/07/2018
* 3845d19 Finaly fixed the nexus issue? - 24/07/2018
* 292d470 Changed Nexus setup - 24/07/2018
* 85a3c3f Moved Nexus home to /home/nexus - 24/07/2018
* 3102e7c Changed to relative paths and versioned Maven - 24/07/2018
* ba27aa1 Updated README - 23/07/2018
* a0beed8 Updated README - 23/07/2018
* fda94a7 Tested other functions - 23/07/2018
* bdeb3a6 Added firewall rule for Jenkins - 23/07/2018
* f3de35b Minor alterations - 23/07/2018
* e6f1d2a Altered the build file - 23/07/2018
* 91a00ab More amendments to fix issues - 23/07/2018
* e42f7ad Amended Java install - 23/07/2018
* 74d4da7 Intitial clean up - 23/07/2018
* dd1f294 Updated README.md - 23/07/2018
* 091bf71 Updated for DocsAppStack - 22/07/2018
* ab594ec Version updates - 22/07/2018
* a4d97fd Added SonarQube and Jenkins plugins - 11/03/2018
* b84e6b1 Removed SDK and added Java and Maven manually - 10032018
* 412d536 Initial commit - 10032018


build.bash
#####################################################################
# Build a Continuous Integration Server for Application Development #
# CentOS v7.4                                                       #
# Dave Collier @10/01/2018                                          #
#####################################################################
# Instructions: Full instruction are in the README.md file          #
# 1. Set the environment variables to your choice in set_env.bash   #
# 2. Run the build script as root: ./build.bash                     #
#                                                                   #
# Contents                                                          #
# --------                                                          #
# Docker / Docker-Compose - To run your containers                  #
# GitBlit - source code repo                                        #
# Jenkins - task runner                                             #
# Maven - build tool and dependecy management                       #
# Nexus - built package and compiled artifact repo                  #
# SonarQube - code analyser                                         #
#####################################################################

# Install OracleJDK
install_java() {
    echo "Installing Java..."
    yum install -y $JAVA_VERSION
}


# Install Apache Maven
install_maven() {
    echo "Installing Maven..."
    mkdir ${MAVEN_DIR}
    cd ${MAVEN_DIR}
    wget http://apache.mirror.anlx.net/maven/maven-3/${MAVEN_VERSION}/binaries/apache-maven-${MAVEN_VERSION}-bin.tar.gz \
    && tar xvfz apache-maven-${MAVEN_VERSION}-bin.tar.gz \
    && ln -s ${MAVEN_DIR}/apache-maven-${MAVEN_VERSION} ${MAVEN_DIR}/mvn3
    echo "export MAVEN_HOME=${MAVEN_DIR}/mvn3" >> ${HOME}/.bashrc
    echo "export PATH=${MAVEN_DIR}/mvn3/bin:$PATH" >> ${HOME}/.bashrc
    source ${HOME}/.bashrc
    sed -e "s|env.MAVEN_REPO|$MAVEN_REPO|g" -e "s|env.NEXUS_USER|$NEXUS_USER|g" -e "s|env.NEXUS_PASS|$NEXUS_PASS|g" -e s"|env.NEXUS_PORT_8081|$NEXUS_PORT_8081|g" < ${BASE_DIR}/settings.xml.orig > ${MAVEN_DIR}/mvn3/conf/settings.xml
    rm -rf ${MAVEN_DIR}/apache-maven-${MAVEN_VERSION}-bin.tar.gz
}

# Set Environment
set_env() {
    echo "Setting environment variables..."
    source ./set_env.bash
}

# Update System
update_system()  {
    echo "Updating system..."
    yum -y update
    yum -y upgrade
    yum install -y yum-utils device-mapper-persistent-data lvm2 wget unzip zip
    yum install -y fontconfig freetype freetype-devel fontconfig-devel libstdc++
}

# Setup Users
setup_users() {
    echo "Setting up Jenkins and Nexus users..."
    adduser -d /home/jenkins -b /bin/bash -u 1000 jenkins
    adduser -d /home/nexus -b /bin/bash -u 200 nexus
    mkdir -p ${DOCKER_VOLUMES}/nexus
    chown nexus:nexus -R ${DOCKER_VOLUMES}/nexus
    mkdir -p ${DOCKER_VOLUMES}/sonarqube/logs
    mkdir -p ${DOCKER_VOLUMES}/sonarqube/conf
    mkdir -p ${DOCKER_VOLUMES}/sonarqube/data
    mkdir -p ${DOCKER_VOLUMES}/sonarqube/extensions
    chmod 777 ${DOCKER_VOLUMES}/sonarqube -R

}

# Install Jenkins
install_jenkins() {
    echo "Installing Jenkins..."
    wget -O /etc/yum.repos.d/jenkins.repo https://pkg.jenkins.io/redhat-stable/jenkins.repo
    rpm --import https://pkg.jenkins.io/redhat-stable/jenkins.io.key
    yum install -y jenkins
    firewall-cmd --zone=public --permanent --add-port=8080/tcp
    systemctl restart firewalld
}

# Docker Install
install_docker() {
    echo "Installing Docker..."
    yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
    yum install -y docker-ce
    systemctl enable docker
    systemctl start docker
    yum install -y epel-release
    yum install -y python-pip
    pip install docker-compose
    pip install --upgrade pip
}

# Docker Compose Install
install_docker_compose() {
    echo "Installing Docker Compose..."
    curl -L https://github.com/docker/compose/releases/download/${DOCKER_COMPOSE}/docker-compose-`uname -s`-`uname -m` -o /usr/local/bin/docker-compose
    chmod +x /usr/local/bin/docker-compose
}

# Build Steps
set_env
update_system
setup_users
install_java
install_maven
install_jenkins
install_docker
install_docker_compose
echo "All Done !!"
# End


set_env.bash
# Choose where you want the base installation to go including the Docker Volumes to go.
export BASE_DIR=/root/DocsAppStack
export DOCKER_VOLUMES=${BASE_DIR}/docker_volumes
export DOCKER_COMPOSE=1.23.2

# GitBlit Setup - Chose the ports
export GITBLIT_PORT_8080=9010
export GITBLIT_PORT_8443=9020
export GITBLIT_PORT_9418=9030
export GITBLIT_PORT_29418=9040

# Nexus/Maven Setup
# The username and password that you will set up Nexus with needs to be exported so that Maven can proxy off the Nexus
# repository to Maven Central to resolve dependencies in the Maven settings file!
export NEXUS_USER=admin    # Default Nexus Username
export NEXUS_PASS=admin123        # Default Nexus Password - change it!
export NEXUS_PORT_8081=9050                     # Nexus webclient port
export MAVEN_VERSION=3.5.4   # Set Maven Version
export MAVEN_DIR=${BASE_DIR}/maven  # Maven installation directory
export MAVEN_REPO=${MAVEN_DIR}/maven-repo       # Default location for the Maven Repository

# SonarQube Setup
export SONARQUBE_PORT_9000=9060
export SONARQUBE_PORT_9092=9070

# Set the docker image versions
export GITBLIT_VERSION=latest
export NEXUS_VERSION=2.14.12
export SONARQUBE_VERSION=7.6-community
export POSTGRES_VERSION=11.2

# Set your version of java
export JAVA_VERSION=java-1.8.0-openjdk-devel
